### ISSUE TITLE
Add rule to require `static propTypes = ..` instad of `Foo.propTypes = ..` for class-based components

### ISSUE THREAD

**Issue by ThiefMaster** [yannickcr/eslint-plugin-react#1923]
With such a rule the following would be invalid:

```javascript
class Foo extends React.Component {}
Foo.propTypes = {...}
Foo.defaultProps = {...}
```

and this would be valid:

```javascript
class Foo extends React.Component {
    static propTypes = {...};
    static defaultProps = {...};
}

function Bar() {
    return <>...</>;
}
Bar.propTypes = {...};
```

**Comment by ljharb** [comment]
Such a rule would need to support both directions: forcing class fields, and forbidding them.

**Comment by ThiefMaster** [comment]
Agreed. Would a blacklist/whitelist make sense? I'd say no, since having e.g. defaultProps outside but propTypes inside the class would make no sense.

**Comment by ljharb** [comment]
I’d think you’d set the rule to “static class field” or “outside the class body”, and then separately disable propTypes, defaultProps, contextTypes, etc (ie, set them to “ignore”).

Also, the rule should forbid (or require as a third option) static getter methods for these things.

**Comment by alexzherdev** [comment]
From what I’ve read here there’s only downsides to using static getters for these, curious why anyone would need to require it?

**Comment by ThiefMaster** [comment]
> If someone isn't able to use class fields yet (they're not yet part of the language)

How likely is it that someone uses react without babel though?

**Comment by alexzherdev** [comment]
I think the point is that not everyone's babel config includes the class fields transform plugin.

**Comment by ljharb** [comment]
@ThiefMaster "with babel" in no way means "with proposals" - eslint can't parse stage 3 features, for example, and using babel-eslint isn't a great solution because it enables parsing of *all* features, not just the ones you're transpiling. Separately, airbnb won't use class fields until there's a lint rule that can forbid ever putting functions in them, for reasons I've gone into elsewhere on this repo.

**Comment by alexzherdev** [comment]
_(last comment off topic)_ 
> Separately, airbnb won't use class fields until there's a lint rule that can forbid ever putting functions in them, for reasons I've gone into elsewhere on this repo.

That sounds like a fairly straightforward rule, any particular reason it hasn't been implemented?

**Comment by ljharb** [comment]
@alexzherdev it'd be long in eslint core, and they don't support anything til it's stage 4.

**Comment by ljharb** [comment]
@kylemh a `static-property-placement` rule ensuring placement of all of: `propTypes`, `defaultProps`, `contextTypes`, `childContextTypes`, `displayName` would be great - the options I'd expect would be "above", "below", "static getters", or "class fields", based on what the ecosystem already does. PRs welcome.

**Comment by alexzherdev** [comment]
I could be wrong, but isn't `sort-comp` responsible for ordering of those?

**Comment by ljharb** [comment]
Ordering, yes - but this is about placement and style (sort-comp doesn’t govern things outside the class, for example)

**Comment by ljharb** [comment]
Closing as a duplicate of #1807.

@dmason30 yes, that's why #1807 has `help wanted` on it.

**Comment by dmason30** [comment]
@ljharb I was more politely asking if you could find the time to do the development. 

I was able to read the labels thanks. :disappointed:

Sorry if I inconvenienced you.

**Comment by ljharb** [comment]
lol ah. if i can find the time, of course! My time however is very limited.
